pkg/metrics/events.go

```diff
@@ -0,0 +1,149 @@
+package metrics
+
+// MetricType will represent the type of the Prometheus metric.
+type MetricType string
+
+const (
+	SampleCountType MetricType = "sample_count"
+
+	ActionSetBackupCreatedType   MetricType = "actionset_backup_created_count"
+	ActionSetBackupCompletedType MetricType = "actionset_backup_completed_count"
+
+	ActionSetRestoreCreatedType   MetricType = "actionset_restore_created_count"
+	ActionSetRestoreCompletedType MetricType = "actionset_restore_completed_count"
+
+	ActionSetTotalCreatedType   MetricType = "actionset_total_created_count"
+	ActionSetTotalCompletedType MetricType = "actionset_total_completed_count"
+)
+
+//MetricTypeOpt is a struct for a Prometheus metric
+type MetricTypeOpt struct {
+	EventFunc  interface{}
+	Help       string
+	LabelNames []string
+}
+
+// Mapping a Prometheus metric name to the metric MetricTypeOpt struct
+var MetricTypeOpts = map[MetricType]MetricTypeOpt{
+	SampleCountType: {
+		EventFunc:  NewSampleCount,
+		Help:       "Sample counter to remove later",
+		LabelNames: []string{"sample"},
+	},
+
+	ActionSetBackupCreatedType: {
+		EventFunc: NewActionSetBackupCreated,
+		Help:      "The count of backup ActionSets created",
+	},
+	ActionSetBackupCompletedType: {
+		EventFunc:  NewActionSetBackupCompleted,
+		Help:       "The count of backup ActionSets completed",
+		LabelNames: []string{"state"},
+	},
+
+	ActionSetRestoreCreatedType: {
+		EventFunc: NewActionSetRestoreCreated,
+		Help:      "The count of restore ActionSets created",
+	},
+	ActionSetRestoreCompletedType: {
+		EventFunc:  NewActionSetRestoreCompleted,
+		Help:       "The count of restore ActionSets completed",
+		LabelNames: []string{"state"},
+	},
+
+	ActionSetTotalCreatedType: {
+		EventFunc:  NewActionSetTotalCreated,
+		Help:       "The count of total ActionSets created",
+		LabelNames: []string{"actionType", "namespace"},
+	},
+	ActionSetTotalCompletedType: {
+		EventFunc:  NewActionSetTotalCompleted,
+		Help:       "The count of total ActionSets completed",
+		LabelNames: []string{"actionType", "namespace", "state"},
+	},
+}
+
+// Event describes an individual event.
+//
+// Note: The type and labels are private in order to force the use of the
```

Comment:
I would like to see the portion of comment block here at top of this file to explain the relationship between MetricType and constructors.

Since we're doing this amount of type checking, we should consider also including a version of the Reflection type checking unit test that Mark wrote for K10. It's a good foundation for Kanister as well.

Reply:
Done.